crypto/x509/x509_test.cc

```diff
@@ -2301,6 +2523,58 @@ TEST(X509Test, Ed25519Sign) {
   ASSERT_TRUE(SignatureRoundTrips(md_ctx.get(), pub.get()));
 }
 
+TEST(X509Test, Dilithium3SignVerifyCert) {
+  //generate the dilithium key
+  EVP_PKEY_CTX *pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  ASSERT_NE(pkey_ctx, nullptr);
+  EVP_PKEY *pkey = EVP_PKEY_new();
+  ASSERT_NE(pkey, nullptr);
+  EXPECT_TRUE(EVP_PKEY_keygen_init(pkey_ctx));
+  EXPECT_TRUE(EVP_PKEY_keygen(pkey_ctx, &pkey));
+
+  //generate the cert
+  bssl::UniquePtr<X509> leaf =
+      MakeTestCert("Intermediate", "Leaf", pkey, /*is_ca=*/false);
+  ASSERT_TRUE(leaf);
+
+  //sign the cert
+  bssl::ScopedEVP_MD_CTX md_ctx;
+  EVP_DigestSignInit(md_ctx.get(), nullptr, nullptr, nullptr, pkey);
+  ASSERT_TRUE(X509_sign_ctx(leaf.get(), md_ctx.get()));
+
+  //verify the cert
+  ASSERT_TRUE(X509_verify(leaf.get(), pkey));
+
+  EVP_PKEY_CTX_free(pkey_ctx);
+  EVP_PKEY_free(pkey);
+}
+
+TEST(X509Test, TestDilithium3) {
+  // This test generates a Dilithium3 certificate from the PEM encoding,
+  // extracts the public key, and then verifies the certificate.
+  bssl::UniquePtr<X509> cert(CertFromPEM(kDilithium3Cert));
+  ASSERT_TRUE(cert);
+
+  bssl::UniquePtr<EVP_PKEY> pkey(X509_get_pubkey(cert.get()));
+  ASSERT_TRUE(pkey);
+
+  ASSERT_TRUE(X509_verify(cert.get(), pkey.get()));
+}
+
+TEST(X509Test, TestBadDilithium3) {
+  // This test generates a Dilithium3 certificate from the PEM encoding
+  // kDilithium3CertNull that has an explicit NULL in the signature algorithm.
+  // After extracting the public key, verifiation should fail.
+  bssl::UniquePtr<X509> cert(CertFromPEM(kDilithium3CertNull));
+  ASSERT_TRUE(cert);
+
+  bssl::UniquePtr<EVP_PKEY> pkey(X509_get_pubkey(cert.get()));
+  ASSERT_TRUE(pkey);
+
+  ASSERT_FALSE(X509_verify(cert.get(), pkey.get()));
+  ERR_clear_error();
```

Comment:
can we check that the right error is returned, like in the `TestEd25519BadParameters` test?

Reply:
added in [8442729](https://github.com/awslabs/aws-lc/pull/759/commits/8442729b7965038b56f2d324e1b6631a36103255)

Reply:
interesting, why does Ed25519 fail with a different error?

Reply:
Looks like `TestEd25519BadParameters` is trying to exercise the `X509_R_INVALID_PARAMETER 136` error from `x509.h` by populating the parameter section with NULL. See [rfc8410](https://www.rfc-editor.org/rfc/rfc8410) section 3 for the following:
> In this document, we define four new OIDs for identifying the
   different curve/algorithm pairs: the curves being curve25519 and
   curve448 and the algorithms being ECDH and EdDSA in pure mode.  For
   all of the OIDs, the parameters MUST be absent.

In the test case I produced, I instead tested the modification of the signatureAlgorithm field to `NULL`, so that a signed certificate without an algorithm can be tested. This is why the error is instead `X509_R_SIGNATURE_ALGORITHM_MISMATCH`.

Reply:
You can verify that this is indeed the case by parsing the ASN.1 encoding for the bad ed25519 certificate and seeing that the value after the OID at offset 31 is `NULL` that represents the `parameters` field of:
```
   AlgorithmIdentifier  ::=  SEQUENCE  {
       algorithm   OBJECT IDENTIFIER,
       parameters  ANY DEFINED BY algorithm OPTIONAL
   }
```

and in my dilithium bad cert we are changing the value of the `signatureAlgorithm` field with a `NULL` inserted at offset 2114.

Reply:
If we wanted to produce the same error as the ED25519 test we could by instead modifying the bytes at offset 31 to add `NULL` parameter selection, as the draft standard for dilithium also proposes that these are absent.